### PR TITLE
Update HPX recipe for HPX V1.8.0

### PR DESCRIPTION
**Pull request by hkaiser**

### CODE REVIEW

File: var/spack/repos/builtin/packages/hpx/package.py [hunk 1]
```diff
@@ -16,23 +16,24 @@ class Hpx(CMakePackage, CudaPackage, ROCmPackage):
...
 
     tags = ['e4s']
 
     version('master', branch='master')
     version('stable', tag='stable')
+    version('1.8.0', sha256='93f147ab7cf0ab4161f37680ea720d3baeb86540a95382f2fb591645b2a9b135')
     version('1.7.1', sha256='008a0335def3c551cba31452eda035d7e914e3e4f77eec679eea070ac71bd83b')
     version('1.7.0', sha256='05099b860410aa5d8a10d6915b1a8818733aa1aa2d5f2b9774730ca7e6de5fac')
-    version('1.6.0', sha256='4ab715613c1e1808edc93451781cc9bc98feec4e422ccd4322858a680f6d9017')
+    version('1.6.0', sha256='4ab715613c1e1808edc93451781cc9bc98feec4e422ccd4322858a680f6d9017', deprecated=True)
```
Comment (msimberg) [inline var/spack/repos/builtin/packages/hpx/package.py]: In what sense is 1.6.0 and older deprecated, but 1.7.X not?

File: var/spack/repos/builtin/packages/hpx/package.py [hunk 2]
```diff
@@ -123,6 +124,15 @@ class Hpx(CMakePackage, CudaPackage, ROCmPackage):
         conflicts("cxxstd=14")
         depends_on("cuda@11:", when="+cuda")
 
+    # Restrictions for 1.8.X
```
Comment (msimberg) [inline var/spack/repos/builtin/packages/hpx/package.py:118]: I can't make a change suggestion on the lines outside of the changed part, but this should look roughly like this: https://github.com/spack/spack/commit/925e5b17fc6a4e1a6d51a2651c7ca0ab0bec1a99. Basically, the asio, boost, and cmake version requirements are the same as 1.7.0 and don't need to be repeated. According to the release notes clang 8 and older is no longer supported. Edit: And the requirements for the latest version are applied to master/release branches, so there's no need to separately have the constraints for those anymore.

**Review by msimberg — CHANGES_REQUESTED** [review]

**Comment by hkaiser** [comment]
@msimberg thanks for the comments, those have been addressed now.

File: var/spack/repos/builtin/packages/hpx/package.py (hunk 2 above)
Comment (msimberg) [reply var/spack/repos/builtin/packages/hpx/package.py:118]: The `with when("@master")` and `with when("@stable")` sections can be removed (see my edit to the previous comment).

Comment (msimberg) [reply var/spack/repos/builtin/packages/hpx/package.py:118]: Likewise `depends_on("cuda@11:", when="+cuda")` needs to be in the 1.8 section.

File: var/spack/repos/builtin/packages/hpx/package.py (hunk 1 above)
Comment (msimberg) [reply var/spack/repos/builtin/packages/hpx/package.py]: I'm still not sure I understand the logic behind which ones are deprecated and which not. Is the idea to deprecate all versions except the latest patch release for each minor release? If yes, shouldn't 1.6.0 stay undeprecated in that case (and 1.1.0 as well)? Or is there some other logic to it?

Comment (hkaiser) [reply var/spack/repos/builtin/packages/hpx/package.py]: Fair points. Apparently, I messed it up. My initial idea was to have only the latest versions of a particular release non-deprecated. I'm not sure what the usual way of marking releases as deprecated is applied for Spack packages.

Comment (msimberg) [reply var/spack/repos/builtin/packages/hpx/package.py]: As far as I understand it `deprecated=True` means that that version is about to removed from spack, not that that version is e.g. no longer supported or something similar upstream.

Comment (hkaiser) [reply var/spack/repos/builtin/packages/hpx/package.py]: Ok, I'll remove the deprecated=True, then. Thanks!

**Review by msimberg — APPROVED** [review]
Thanks @hkaiser!

**Review by tldahlgren — APPROVED** [review]
Confirmed sha256.

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Pull request merged**